keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,415 @@
+   be utilized as knowledge sharing with cloud providers.
+
+## Motivation
+
+The motivation for each change is:
+
+1. Nodes used as intermediate nexthop for `eTP:Cluster` services would allow all
+  connections passing through the node while it is being terminated, to
+  gracefully shutdown.
+
+2. Cloud providers have very different ways of ascertaining if a load balancer
+   should target a specific node for `eTP:Cluster` services. We would like to
+   highlight the benefits of certain methods and pitfalls of others, in a formal
+   document, so that this is known. The hope is that this allows the information
+   to act as a source of knowledge for how to adapt their implementations to the
+   mechanics of a Kubernetes cluster.
+
+### Goals
+
+- Offer a better capability of connection draining terminating Nodes, for load
+  balancers which support that.
+
+### Non-Goals
+
+- Aligning cloud provider HCs for `eTP:Cluster` services. Cloud providers like
+  Azure/AWS do not configure their HCs to point to the service proxy. Instead
+  they connect to the `NodePort` defined for the service. As to have them
+  benefit from the proposals of this KEP, they would need to change their
+  implementation.
+
+- That Kube-proxy includes its `healthz` state AND its current answer w.r.t the
+  local endpoints, when it answers to the HC for `eTP:Local` services.
+  Kube-proxy is currently defined as "unhealthy" when `2 * syncPeriod` passes in
+  which it knows that it needs to update the data plane (iptables/ipvs), but has
+  not actually done so. Not including the `healthz` state can cause Kube-proxy
+  to indicate to a load balancer that it should send traffic to a Node simply
+  because the endpoint is scheduled there, even though Kube-proxy might not be
+  healthy and successfully managed to write the rules required for actually
+  being able to forward traffic to the endpoint. This has however been agreed is
+  a bug and will be treated as such, as opposed to following the KEP cycle for
+  it.
+
+## Proposal
+
+#### Risk
+
+The risk are:
+
+1. Vendors of Kubernetes which deploy Kube-proxy and specify a `livenessProbe`
+   targeting `/healthz` are expected to start seeing a CrashLooping Kube-proxy
+   when the Node gets tainted with `ToBeDeletedByClusterAutoscaler`. This is
+   because: if we modify `/healthz` to fail when this taint gets added on the
+   Node, then the `livenessProbe` will fail, causing the Kubelet to restart the
+   Pod until the Node is deleted. No vendor has been found which does this
+   currently.
+
+2. By not being able to watch the Node object (while failing to read from the
+   API server, for example) we might have all Kube-proxy start failing the HCs
+   at once. That being said: Kube-proxy currently watches the Node object and is
+   susceptible to this risk.
+
+#### Mitigations
+
+1. Such problems are expected to surface during the Beta phase when the feature
```

Comment:
Also: As far as we can tell, none of the major cloud providers set `livenessProbe`, nor does kubeadm, so the risk is low.

Reply:
Mentioned the fact that the risk is low in `Risk`